contracts/near/eth-client/src/lib.rs

```diff
@@ -348,12 +383,192 @@ impl EthClient {
+
+        if !is_epoch && signers_bytes != 0{
+            return false;
+        }
+
+        if is_epoch && signers_bytes % validator_bytes_length != 0{
+            return false;
+        }
+
+        // Ensure that the mix digest is zero as we don't have fork protection currently
+        if header.mix_hash != H256([0; 32].into()) {
+            return false;
+        }
+
+        // Ensure that the block doesn't contain any uncles which are meaningless in PoA
+        if header.uncles_hash == H256([0; 32].into()) {
+            return false;
+        }
+
+        return self.verify_cascading_fields(header, prev);
+    }
+
+    fn verify_cascading_fields(&self, header: &BlockHeader, prev: &BlockHeader) -> bool {
+        
+        // The genesis block is the always valid dead-end
+        if header.number == 0 {
+            return true;
+        }
+        
+        // check if the parent hash and bumber match.
+        if prev.number != header.number - 1 || prev.hash.unwrap() != header.parent_hash {
+            return false;
+        }
+
+        let capacity = 0x7fffffffffffffff as u64;
+        let (gas_limit_bound_divisor, min_gas_limit) = (256, 5000);
+        
+        // Verify that the gas limit is <= 2^63-1
+        if header.gas_limit > U256(capacity.into()) {
+            return false;
+        }
+        
+        // Verify that the gasUsed is <= gasLimit
+        if header.gas_used > header.gas_limit {
+            return false;
+        }
+        
+        let prev_gas_limit = format!("{}", prev.gas_limit).parse::<i64>().unwrap();
+        let header_gas_limit = format!("{}", header.gas_limit).parse::<i64>().unwrap();
+
+        // Verify that the gas limit remains within allowed bounds
+        let diff = (prev_gas_limit-header_gas_limit).abs();
+        let limit = prev_gas_limit / gas_limit_bound_divisor;
+
+        if diff >= limit || header.gas_limit < U256(min_gas_limit.into()) {
+            return false;
+        }
+        
+        self.verify_seal(header)
+    }
+    
+    // check if the block is an epoch header.
+    fn is_epoch(&self, number: u64)->bool{
+        number%200==0
+    }
+    
+    // check if the author is the signer.
+    fn is_author(&self, header: &BlockHeader) -> bool{
+        
+        let extra_seal = 65;
+        let seal_hash = self.seal_hash(header, U256(self.chain_id.into()));
+
+        // get the signature from header extra_data
+        let signature = header.extra_data[header.extra_data.len()-extra_seal..].to_vec();
+        let mut sig = [0u8; 65];
+        sig.copy_from_slice(&signature[..]);
+
+        let v = sig[64];
+        let mut r = [0u8; 32];
+        let mut s = [0u8; 32];
+        r.copy_from_slice(&signature[0..32]);
+        s.copy_from_slice(&signature[32..64]);
+        
+        let rec_id = RecoveryId::parse(v).unwrap();
+        let mut data = [0u8; 64];
+        data[0..32].copy_from_slice(&r[..]);
+        data[32..64].copy_from_slice(&s[..]);
```

Comment:
What is the difference between `signature` and `data`?

Reply:
the signature is 65 bytes and data 64